### PR TITLE
Scheduled Downtime Notifications

### PR DESCRIPTION
**Pull request by ncksllvn**
This PR implements scheduled downtime throughout our website for application that are all or nothing - if an external service that an application depends on is unavailable, then so is the application. These applications include:

- ~~The claim status tool~~
- The health records tool
- The secure messaging tool
- The Veteran ID Card application
- The prescription refill tool
- The facility locator
- The user profile page
- The Post-9/11 GI Bill benefits tracking tool
- ~~The  appeals status tool~~

Below are the different states of an application that implements Downtime Notifications.

# Status: OK
Application is rendered as normal.

---------------------
# Status: Downtime Approaching (< 1 hour)

image

---------------------
# Status: In Downtime

image

----------------------

Resolves department-of-veterans-affairs/vets.gov-team/issues/7113
[Spreadsheet about application dependencies](https://docs.google.com/spreadsheets/d/1QtHiKZW5DHpm8oZ8rtNFxCfgVNyIW7UNqaeBDdQg8u4/edit#gid=0)

### CODE REVIEW

File: src/js/common/helpers/api.js
Comment (ncksllvn) [inline src/js/common/helpers/api.js]: I'm adding a check here because otherwise if sessionStorage.userToken is `undefined`, then the authorization header is set to `Token token=undefined`. Then our mock API registers the requests to that instead of `_global`.

File: test/common/components/RequiredTermsAcceptance.unit.spec.jsx
Comment (ncksllvn) [inline test/common/components/RequiredTermsAcceptance.unit.spec.jsx]: I have no idea why this test was failing. I don't think it has anything to do with this PR because it also failed when I switched to master and reran it there locally...so I think another test was initializing `window.dataLayer` and this test would usually run after that.

**Comment by ncksllvn** [comment]
#### Design Implementation
An API route provides an array of downtime windows pertaining to different services. A React components called `DowntimeNotifications` can then be used to consume that information by passing it a property called `dependencies`', which is a list of external services for which that application depends. The JSX returned from the `render` function can then be wrapped in a `<DowntimeNotifcation/>`, and if the API provides information about an external service going offline within the hour or is current offline, your application will respond accordingly.

So, the code for some app is something like below. In the case that MHV is going offline within the hour, the user is told that information but some app is still usable. If MHV is currently is offline, then the user is told that information and they are not shown some app.

```jsx
import React from 'react';
import DowntimeNotification, { services } from '../../common/containers/DowntimeNotification';

class SomeApp extends React.Component {
    render() {
        return (
            <DowntimeNotification dependencies={[services.mhv]}>
                <h1>Hey there, if you're reading this, then MHV must be good to go!</h1>
            </DowntimeNotification>
        )
    } 
}
```

**Comment by ncksllvn** [comment]
To mock data for testing, you can see this [commit ](https://github.com/department-of-veterans-affairs/vets-website/blob/91834c9c09eeefe003eb336755ed845de4f06a1f/src/js/common/actions/index.js) where instead of an API hit the Redux action just supplies test data. Or you can modify your local API's maintenance_window_controller.rb to do something like below. Or actually throw some data in your local database and do it that way.

```ruby
   def index
      render json: { 
        data: [
        {
          id: '139',
          type: 'maintenance_windows',
          attributes: {
            external_service: 'appeals',
            description: 'Benefits services will be unavailable for 24 hours.',
            start_time: Time.zone.now.yesterday,
            end_time: Time.zone.now.tomorrow,
            description: 'maintenance from yesterday'
        }
       ]
     }
```

**Review by melwoodard — DISMISSED** [review]
Hi @ncksllvn -- I think there was some miscommunication on what the UI should be for this first iteration. The UI should look the same whether authenticated or not, the only difference is whether downtime is _approaching_ vs. _in effect_ 

Intended presentation when the user lands on the affected application:

1. Downtime approaching = modal overlays the page

2. Downtime in effect = plain messaging replaces all content below the page title

It should just be a matter of swapping the two alert banners you show above to instead be an overlay (for upcoming) and plain text (for in effect).

**Comment by ncksllvn** [comment]
Thanks @melwoodard, I guess I overlooked that change in requirements. I think we will need to add another ticket for when we want to add the functionality back in later because I didn't include it in Pt 2 or 3.

I'll make the changes asap.

**Comment by melwoodard** [comment]
Thanks @ncksllvn - to be clear, the different experience for auth vs. non-auth is what's outlined in the Pt 3 ticket (https://github.com/department-of-veterans-affairs/vets.gov-team/issues/8124) -- showing an alert in the sign in modal when a non-authed user tries to access an auth-required app. So we should be good!

**Comment by ncksllvn** [comment]
@melwoodard Ohhh, I was under the impression that that ticket was specifically for warnings in the Sign-in Modal because that requires the extra step of mapping the URL parameters to application dependencies. Are we planning on ever adding back the stuff that was originally in this PR that changed used the warning banners?

I just updated this PR with those updates btw.

**Comment by melwoodard** [comment]
Nope, the warnings in the sign in modal cover the need, we don't need the warnings you initially had in this PR.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [inline src/js/common/containers/DowntimeNotification.jsx]: So if we have something like
```js
downtimes = [
  { startTime: '10:00', endTime: '11:00' },
  { startTime: '12:00', endTime: '13:00' }
}
```
then `getDowntimeWindow(downtimes)` will return `{ startTime: '11:00', endTime: '13:00' }`, is that right?
(This took me entirely too long to figure out what the purpose was. :grimacing: )

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [inline src/js/common/containers/DowntimeNotification.jsx]: Perhaps an example would help clear this description up? Could just be me, but I was having a hard time making sense of it.

File: src/js/common/helpers/api.js
Comment (ncksllvn) [inline src/js/common/helpers/api.js]: The way this was before we would see `Token token=undefined` and that didn't work well with the mock API

**Comment by ncksllvn** [comment]
Thanks @melwoodard, sounds good! Sorry about the mixup there.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [inline src/js/common/containers/DowntimeNotification.jsx]: Ooh, didn't know `LT` was a thing. Very cool.
Just to make sure all our bases are covered, have you made sure this plays nice with time zones? I know Mark ran into some funny stuff a few days ago with time zones.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [inline src/js/common/containers/DowntimeNotification.jsx]: Couple things:
1. Straight quote needs to be a curly one (I'm not sure how to do this on this linux machine, so no example, sorry!)
2. Looks like there's an extra space between `{endTime.format('LT')}` and the following `.`

File: src/js/claims-status/containers/ClaimsStatusApp.jsx
Comment (ncksllvn) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: The former was my thinking too, but they'll actually be asked to login before they land on the application anyhow because of the way the menu links works with the `?next=` redirect. We have some work planned for a third piece that will tell the user in the login window prior to logging in if the application is experiencing downtime but for now it only checks when landing on the application. 

If they navigate directly (via bookmark or something) order would matter though - do you think it's worth making the change?

Comment (ncksllvn) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: Originally we were planning on difference messaging based on authenticated or not authenticated, so the downtime notifications had to wait anyway.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: No great explanation other than that I really didn't want to add a rule for just that single margin. Think it's worth adding?

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: It did work but it failed on Jenkins somehow!

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Right now it calculates downtime based on landing and that's all. That's how we defined it in our requirements to keep things simple in this first iteration. At a later date we may figure out those "transitional" states.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: I wrote this assuming that outages will be at least an hour long so I liked the explicitness of `isBetween` but otherwise I think you're right. @patrickvinograd can you confirm?

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Yeah exactly! It has the full downtime object as returned from the action [here](http://jenkins.vetsgov-internal/blue/organizations/jenkins/testing%2Fvets-website/detail/scheduled_downtime/13/pipeline), but it only needs that properties. This may have been overkill, but it's ultimately finding the earliest start time and the latest end time from the application's dependencies and returning that as the downtime window.

Is this too cryptic? I tried to write it clearly but it was one of those things were it doesn't seem like there was a super clear way to do it.

Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Sort of - it's ultimately just trying to find the earliest start time and the latest end time. This is probably overkill but I figured it's also the most correct way to do it in case there's more than one outage. I might rewrite this with variables named more explicitly for `earliestStart`.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Good idea, will do

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [reply src/js/common/containers/DowntimeNotification.jsx]: Ohh, that's interesting...Failed in the unit test, which wouldn't have loaded up the `src/js/common/polyfill.js`...but that should be running in Node which I thought shouldn't have a problem with `Object.values`. Unless we're running an old version?

Well, I guess we can ignore it for now. That's so odd, though!

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: I believe it should be adjusted for the [locale](https://momentjs.com/docs/#/displaying/)

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [reply src/js/common/containers/DowntimeNotification.jsx]: Oh, I get it: [Mozilla](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values) says it's ES2017, which isn't ES6 (ES2015)...which [isn't supported](http://node.green/#ES2017-features-Object-static-methods) in Node v6.11.1 (and _that's_ what Jenkins is running the unit tests in).

ES2017 !== ES6 :man_facepalming: 

My bad.

File: src/js/claims-status/containers/AppealStatusPage.jsx
Comment (jbalboni) [inline src/js/claims-status/containers/AppealStatusPage.jsx]: On this app, this is only going to show up for the downtime approaching case, correct? Because if appeals is down, they'll be stuck on the claim list page.

File: src/js/claims-status/containers/ClaimsStatusApp.jsx
Comment (jbalboni) [inline src/js/claims-status/containers/ClaimsStatusApp.jsx]: This is going to block someone from viewing their appeals if evss is down. Is that ok?

File: src/js/common/containers/DowntimeNotification.jsx
Comment (jbalboni) [inline src/js/common/containers/DowntimeNotification.jsx]: I doubt this one will ever be used. For edu, we submit over SFTP and going down doesn't impact users submitting applications.

**Comment by ncksllvn** [comment]
Thanks for the CR, @cvalarida! Will make some updates here soon.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (jbalboni) [inline src/js/common/containers/DowntimeNotification.jsx]: I think you could use the native `Map` for this.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [reply src/js/common/containers/DowntimeNotification.jsx]: Yeah, I'm not sure. Is it in a preexisting pattern that may get a style update later? If so, it might be worth it. If not, I wouldn't worry about it.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (jbalboni) [reply src/js/common/containers/DowntimeNotification.jsx]: Yeah, I think in the case of the longer forms, the downtime pattern will have to be different. We would probably want to notice logged in users in place of the submit button, and notify non-logged in users some less invasive way. Definitely makes sense to leave for the second iteration.

File: src/js/claims-status/containers/ClaimsStatusApp.jsx
Comment (cvalarida) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: Ahh, yeah that makes sense. I wouldn't worry about it.

At first, I didn't think the redirects happened until after the page was loaded up, but then I found this little number in `src/js/login/containers/Main.jsx`

```js
  bindNavbarLinks() {
    [...document.querySelectorAll('.login-required')].forEach(el => {
      el.addEventListener('click', e => {
        e.preventDefault();
        const nextQuery = { next: el.getAttribute('href') };
        const nextPath = appendQuery('/', nextQuery);
        history.pushState({}, el.textContent, nextPath);
        this.props.toggleLoginModal(true);
      });
    });
  }
```

I don't like :point_up: because it's not obvious that this is happening, but it's there and out of scope for this PR, so...I would say don't worry about it. The way you've got it right now makes enough sense. Navigating directly to an app with a required login is probably not super common (though I would love to see analytics on that).

File: src/js/common/containers/DowntimeNotification.jsx
Comment (jbalboni) [reply src/js/common/containers/DowntimeNotification.jsx]: There's also a particular way that the content team has defined for showing AM and PM, that we've had to set directly: https://github.com/department-of-veterans-affairs/vets-website/blob/933acc5b11ab52db6269fcc0ae275cb6b33a06b7/src/js/common/schemaform/save-in-progress/RoutedSavableApp.jsx#L31. So may want that in here, too.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (jbalboni) [inline src/js/common/containers/DowntimeNotification.jsx]: This would be clearer, I think, if it weren't a one line if statement.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [reply src/js/common/containers/DowntimeNotification.jsx]: Do we know what we'd expect to get from the api if we had downtimes on the same day, but like 10 hours apart? Would it give us both? I'm wondering if this means it'll say we've got a downtime of 11 hours, when we really only have two 1-hour downtimes.

File: src/js/claims-status/containers/ClaimsStatusApp.jsx
Comment (ncksllvn) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: Yeah exactly! I definitely agree. That's something we might be working with later on.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: I think it's okay at the moment. If any more style gets added then it'll definitely make sense to move it over but I think it might be staying pretty simple. Hoping to get by using default components/tags.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Haha no worries, this is deep into the web of JS frontend/backend and preprocessors. We may need to revisit this at some point.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: That is a really great point, nice find. I suppose the pattern it should follow is finding the earliest start point, and the end point based on continuous outage...or maybe that's overkill and it should just be earliest start point and just use the corresponding end point.

Comment (patrickvinograd) [reply src/js/common/containers/DowntimeNotification.jsx]: I vote for the latter - just display the next upcoming outage for a given service. I doubt we'll have windows defined in rapid succession or overlapping, but we might, say, populate multiple distinct windows coming up in the next week when we hear about them.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Good suggestion, thanks!

File: src/js/claims-status/containers/AppealStatusPage.jsx
Comment (ncksllvn) [reply src/js/claims-status/containers/AppealStatusPage.jsx]: It was implemented based purely off of the spreadsheet; I don't know enough about the application to know to do otherwise. Is this a special case? If it is we can hold off for later.

**Comment by ncksllvn** [comment]
@jbalboni Definitely let us know if you have some insight into the way Appeals is structured with its dependencies. If it isn't as simple as suggested in the [spreadsheet](https://docs.google.com/spreadsheets/d/1QtHiKZW5DHpm8oZ8rtNFxCfgVNyIW7UNqaeBDdQg8u4/edit#gid=0) we'll hold off for a later date.

File: src/js/claims-status/containers/ClaimsStatusApp.jsx
Comment (ncksllvn) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: Not sure, is there a better place to add the Downtime Notification? I went purely based on the spreadsheet and what seemed to me as the best entry points but feel free to point me in the right direction if I was off on any of this.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: That sounds good to me!

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Yeah good call. At one point it was nice and clean as a one liner but not so much now.

File: src/js/claims-status/containers/ClaimsStatusApp.jsx
Comment (jbalboni) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: This app has two independent dependencies, so given the current design, I don't think there's a good place to put it that doesn't block use of the app if just one of the dependencies is down.

Comment (ncksllvn) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: The downtime notification doesn't have to wrap the whole thing - is it blocking a menu in this case? And that's why they wouldn't be able to view their appeals if evss is down?

Comment (jbalboni) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: In the claim status app the initial page of the app is a list of claims and appeals, mixed together. My understanding is that this would replace that whole page with the downtime message.

Comment (cvalarida) [reply src/js/claims-status/containers/ClaimsStatusApp.jsx]: Yeah, this is one of those places that an all-or-nothing block isn't a good fit. @ncksllvn here's kind of how it works:

The "claims-status" app is a bit of a misnomer. It acts as a way for the Veteran to view their claims _and_ the appeals to those claims (which are often seen as part of the claims themselves, which is why we've kept the name). When we go to the app, it'll query for both claims and appeals and list them on the index page. After one is selected, it'll open up either the claims status page or the appeals status page, depending on the type.

The trouble here is that claims and appeals rely on two different services. We could block those status pages since they rely on one service each, but that doesn't give the user any warning that a maintenance window has been scheduled. Ideally, we'd be able to handle showing messages to indicate either claims or appeals are down, and block the whole app if they're both down.

But that's not MVP.

@djmunoz1 or @rroueche, does it make sense to have this downtime notification block the app if _either_ evss or the appeals service is down?

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [reply src/js/common/containers/DowntimeNotification.jsx]: Idle curiosity: What's the benefit of using a native Map vs just an object?

File: src/js/claims-status/containers/AppealStatusPage.jsx
Comment (cvalarida) [reply src/js/claims-status/containers/AppealStatusPage.jsx]: Like I mentioned in another comment, this page is basically only accessed from the claims & appeals index page. If the appeals service is down, the appeals just won't show up in the index page, so the user won't be able to navigate here. Still, if they have it bookmarked or something, this'll help. What we do here will depend on what we do with the `ClaimsStatusApp`.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (jbalboni) [reply src/js/common/containers/DowntimeNotification.jsx]: Clarity, mainly. Map is the data type that is intented for this use case. It is also slightly faster in some uses cases, but I doubt enough to matter here.

**Comment by ncksllvn** [comment]
@jbalboni should we consider moving that `moment.updateLocale('en', { ...` stuff that the content team prefers into a single file somewhere? I see it several times in our project.

**Comment by ncksllvn** [comment]
@jbalboni  @cvalarida  We moved Appeals and Claims to iteration 2 to keep things simple. If you guys don't mind taking another look, I think all the fixes have been made. Thanks a ton for your awesome reviews, definitely appreciated. 

We're trying to finish QA'ing with the Pager Duty API tomorrow and go live if possible so please keep me posted if you have any questions/concerns. If you want to run locally and need help getting set up I'm happy to help.

@leannammiller for visibility

File: src/js/common/containers/DowntimeNotification.jsx
Comment (cvalarida) [inline src/js/common/containers/DowntimeNotification.jsx]: Maybe splitting hairs here because this may never happen, but if we've got the following downtimes
```js
downtimes = [
  { startTime: '12:00', endTime: '13:00' },
  {  endTime: '15:00' }
]
```
I think this will return
```js
{
  startTime: moment(undefined), // resolves to the current time
  endTime: moment('15:00')
}
```
when the current time is < 12:00. This might be the intended behavior, but I wanted to point it out.

File: test/common/containers/DowntimeNotification.unit.spec.jsx
Comment (cvalarida) [inline test/common/containers/DowntimeNotification.unit.spec.jsx]: Huh. Well, if this passes, I'm satisfied. When I was trying to figure out how `getDowntimeWindow` might break things, I found if `downtimes` was an empty array, it should return `{ startTime: false, endTime: false }` which would cause an error in `renderStatusDown` or `renderStatusDownApproaching`. It also seemed like `getDowntimeWindow` _could_ be called with an empty array from `componentWillRecieveProps`.

Turns out I must be missing something, though, if this passes. :man_shrugging:

**Review by cvalarida — APPROVED** [review]
Looks good from where I sit! I was concerned that it might break, but if the tests pass, I think it's golden.

File: src/js/common/containers/DowntimeNotification.jsx
Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Actually it's only looking at the `startTime` now and grabbing the corresponding `endTime` off that. So that should return 

```
{
  startTime: moment('12:00'), // resolves to the current time
  endTime: moment('13:00')
}
```
But that brings up a good point, because the endTime can be omitted in which case we shouldn't pass it into moment. I just published a patch for that case.

Comment (ncksllvn) [reply src/js/common/containers/DowntimeNotification.jsx]: Oh I see what you mean. The `startTime` is required though, so that shouldn't ever happen. `endTime` though is optional, which isn't clear here. I just published a patch for this.

File: test/common/containers/DowntimeNotification.unit.spec.jsx
Comment (ncksllvn) [reply test/common/containers/DowntimeNotification.unit.spec.jsx]: It's confusing, but I just added a check to do the `getDowntimeWindow` only when the status isn't ok. There was already a check for that prior to this, but it wasn't clear enough. It was tough working with so many date operations.

**Comment by DanielleSoCompany** [comment]
@leannammiller let me know if this works for the new cases:

**Please save your application and come back to submit it after [DOWNTIME DATE]

We’re making some updates to this application right now. You can still fill out the application. And, if you’re signed in to Vets.gov, you can save your information. But you won’t be able to submit your application until we’re finished with our updates. Please save your application and come back after [DOWNTIME DATE] to submit it.

**We won’t be able to accept applications from [DOWNTIME START DATE] to [DOWNTIME END DATE]**

We’re going to be making some updates to this application from [DOWNTIME START DATE] to [DOWNTIME END DATE].** During this time, you’ll still be able to fill out the application. And, if you’re signed in to Vets.gov, you’ll be able to save your information. But you won’t be able to submit your application until we’re finished with our updates. Please save your application and come back after [DOWNTIME DATE] to submit it.

**Comment by jkassemi** [comment]
Aside: this would be great to tie into the future user notifications feature.

**Comment by melwoodard** [comment]
@jkassemi read my mind -- I'm discussing with @Samara-Strauss to figure out how this fits in with asynchronous submission.

**Comment by Samara-Strauss** [comment]
Noted! We're actually talking about how to handle downtime here https://github.com/department-of-veterans-affairs/vets.gov-team/issues/8573#issuecomment-364206744

**Pull request merged**
